Add a flag to start processing statediff if we are caught up to the head of the chain

abdulrabbani00 opened the pull request:
# Overview
This PR will allow users to add a flag when starting geth, indicating if we should wait for the statediff service to catch up to the head of the chain. 

* The feature has been tested using dapptools.
* The feature includes unit tests.
* The feature has __not__ been tested with mainnet data.
* The feature has __not__ been tested for performance.

# Data Structure Changes

This feature changes two major data structures used in the statediffing service.
1. `statediff/config.go` - `Config` 
  a. `WaitforSync` - This change allows us to indicate if the statediff service should wait for us to catch up to head.
2.  `statediff/service.go` - `Service`
  a. `WaitforSync` - This change allows us to indicate if the statediff service should wait for us to catch up to head.
  b. `Backend` - This backend object is necessary to query the internal API, which indicates if we have caught up to the head of the chain.

These changes have not failed any of the current unit tests under `/statediff.`

# Testing

This PR includes two unit tests. These unit tests have a new mock called `Backend.` The tests make sure that we catch up to head properly before indicating that we are all caught up.

We should test this with production data to ensure the feature is working correctly and does not incur any performance issues.

Commit:
Add a flag to start processing statediff if we are caught up to the head of the chain

Commit:
Update param --> parameter to fix conflicting names

Commit:
Fix lint errors

i-norden commented on statediff/README.md:
```diff
@@ -71,38 +73,38 @@ type Payload struct {
-`--statediff.db.nodeid` is the node id to use in the Postgres database  
-`--statediff.db.clientname` is the client name to use in the Postgres database  
-`--statediff.file.path` full path (including filename) to write statediff data out to when operating in file mode  
+`--statediff` flag is used to turn on the service
```
Without the 2 spaces after each line these show up all on the same line when viewing the readme in github.

abdulrabbani00 replied:
Thank you, not sure why my markdown formatted disrespected me like this.

i-norden commented on statediff/service.go:
```diff
@@ -528,10 +535,48 @@ func (sds *Service) Unsubscribe(id rpc.ID) error {
+// It will return false if geth has finished syncing.
+// It will return a non false value if geth is not done syncing.
+func (sds *Service) GetSyncStatus(pubEthAPI *ethapi.PublicEthereumAPI) (interface{}, error) {
+	syncStatus, err := pubEthAPI.Syncing()
```
We can just return `pubEthAPI.Syncing()` directly unless we want to wrap additional context around the `err` or we want to resolve the `interface{}` to True or False and return `(bool, error)` instead.

abdulrabbani00 replied:
I've made two changes:

1. I handle the error in the `Start` function. The error from `GetSyncStatus` gets passed all the way to the top.
2. I return a bool instead of an `interface`. At this point, the interface is not useful, but there could be a day when the information in the interface is useful (it provides a rich map of the state of the chain when the sync occurred).

i-norden commented on statediff/service.go:
```diff
@@ -528,10 +535,48 @@ func (sds *Service) Unsubscribe(id rpc.ID) error {
+	// Has the geth node synced to head?
+	Synced := false
+	pubEthAPI := ethapi.NewPublicEthereumAPI(sds.BackendAPI)
+	for !Synced {
```
We should use a `time.Ticker` here to only poll at some short interval like 1 second.

i-norden commented on statediff/service_test.go:
```diff
@@ -291,3 +291,151 @@ func testErrorInStateDiffAt(t *testing.T) {
+		}
+
+		// This code will only be run if the sync is complete and the WaitForSync function is complete
+		//t.Log("Backend: ", backend)
```
nitpick: clean up these commented out loggings

i-norden commented on statediff/service_test.go:
```diff
@@ -291,3 +291,151 @@ func testErrorInStateDiffAt(t *testing.T) {
+			}
+
+		} else {
+			t.Error("checkSyncComplete is empty: ", len(checkSyncComplete))
```
nitpick: I think the only way it'd reach this code path is if `len(checkSyncComplete)` was somehow > 1, not empty (due to the 3 guards starting on line 403). Which is to say idt this code path is possible, so not sure we need the `if len(checkSyncComplete) == 1` check on line 427.

abdulrabbani00 replied:
Good catch there. Sharp eye 👁️

i-norden commented on statediff/service_test.go:
```diff
@@ -291,3 +291,151 @@ func testErrorInStateDiffAt(t *testing.T) {
+// It will then incrementally increase the currentBlock to match the highestBlock
+// At each interval it will run the GetSyncStatus to ensure that the return value is not false.
+// It will also check to make sure that the WaitForSync() function has not completed!
+func testGetSyncStatus(t *testing.T) {
```
Nice, this unit test is very well formulated/structured!

abdulrabbani00 replied:
Thank you sir :D

i-norden reviewed (changes requested):
Looks great! Just a few comments for some very minor non-functional changes.

Commit:
Updates as Per Ian's comments

1. Fix some naming inconsistencies.
2. `GetSyncStatus` returns a bool
3. `Start` handles any `errors` from the `WaitingForSync` function.
4. Add a sleep for the `while` loop.
5. Fix test based on a conditional that would never be met.

i-norden reviewed (approved):
LGTM! Thanks for making the updates.
